core/src/main/java/org/kohsuke/stapler/export/Property.java

```diff
@@ -199,118 +197,122 @@ private void writeValue(Type expected, Object value, TreePruner pruner, DataWrit
 
         Class c = value.getClass();
 
-        Model model;
-        try {
-            model = owner.get(c, parent.type, name);
-        } catch (NotExportableException ex) {
-            if(STRING_TYPES.contains(c)) {
-                writer.value(value.toString());
-                return;
+        if (c.getAnnotation(ExportedBean.class) == null) {
+            handleNotExportable(expected, value, pruner, writer, skipIfFail, c);
+        } else {
+            try {
+                writer.type(expected, value.getClass());
+            } catch (AbstractMethodError _) {
+                // legacy impl that doesn't understand it
             }
-            if(PRIMITIVE_TYPES.contains(c)) {
-                writer.valuePrimitive(value);
-                return;
+            Model model = owner.getOrNull(c, parent.type, name);
+            if (model == null) {
+                handleNotExportable(expected, value, pruner, writer, skipIfFail, c);
+            } else {
+                writer.startObject();
+                model.writeNestedObjectTo(value, pruner, writer);
+                writer.endObject();
             }
-            Class act = c.getComponentType();
-            if (act !=null) { // array
-                Range r = pruner.getRange();
-                writer.startArray();
-                if (value instanceof Object[]) {
-                    // typical case
-                    for (Object item : r.apply((Object[]) value)) {
-                        writeBuffered(act, item, pruner, writer);
-                    }
-                } else {
-                    // more generic case
-                    int len = Math.min(r.max, Array.getLength(value));
-                    for (int i=r.min; i<len; i++) {
-                        writeBuffered(act, Array.get(value, i), pruner, writer);
-                    }
+        }
+    }
+
+    private void handleNotExportable(Type expected, Object value, TreePruner pruner, DataWriter writer, boolean skipIfFail, Class c) throws IOException {
```

Comment:
The method name is potentially misleading. Really this method is handling anything without `@ExportedBean`, but that includes all sorts of objects which are “exportable”. It is only at the very end, starting where we check `skipIfFail`, that this method is in fact handling objects which are not exportable.

Reply:
Renamed and added a description.